src/sidebar/components/annotation-document-info.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const propTypes = require('prop-types');
+const { createElement } = require('preact');
+
+const annotationMetadata = require('../annotation-metadata');
+const memoize = require('../util/memoize');
+
+/**
+ * Render some metadata about an annotation's document and link to it
+ * if a link is available.
+ */
+function AnnotationDocumentInfo({ annotation }) {
+  const docMetaMemoFn = memoize(annotationMetadata.domainAndTitle);
```

Comment:
Is the general plan that we'd do away with the `memoize` util after migration, or does it have use for non-UI layers of the app indefinitely?

Reply:
Regarding the use of `memoize` here, I was porting the Angular variant a little thoughtlessly. I'd imagine there must have been some sort of re-render issue with the document info (this hunch is made stronger by this line in the old template https://github.com/hypothesis/client/blob/d7a959c1ecdf9558a0839b9b3f355336b976a25e/src/sidebar/templates/annotation-header.html#L22 which suggests that someone was trying to squelch something). 

Unless there is some reason that document metadata would be flailing-ly updating on `annotation` objects without a meaningful change to the annotation itself, it doesn't seem like we should _need_ to memo-ize this here.

Reply:
> Unless there is some reason that document metadata would be flailing-ly updating on annotation objects without a meaningful change to the annotation itself, it doesn't seem like we should need to memo-ize this here.

It is true that memoization in React/Preact is only an optimization, rather than being a requirement in some cases with Angular JS to cause its change detection loop to terminate. From a quick skim of the functions in `annotation-metadata.js` it looks like they do enough work that it would probably still be worth doing using `useMemo`. That's not essential to merge this though.

Something to be aware of is that React/Preact don't prevent skip re-rendering a component by default if the inputs haven't changed, they just avoid updating parts of the DOM where the output hasn't changed since the last render. I haven't done any profiling yet but I'm sure we'll  end up making use of [some of the methods available](https://reactjs.org/docs/optimizing-performance.html#avoid-reconciliation) to skip-rerendering of certain parts of the UI tree if inputs haven't changed.